storage/aptosdb/src/lib.rs

```diff
@@ -265,7 +264,8 @@ pub struct AptosDB {
     system_store: Arc<SystemStore>,
     transaction_store: Arc<TransactionStore>,
     pruner_config: StoragePrunerConfig,
-    pruner: Option<Pruner>,
+    state_pruner: Arc<Option<Box<dyn PrunerManager + Send + Sync>>>,
```

Comment:
1. `Option<StatePrunerManager>` should do?
2. Not introduced by this, but: imagine when we disable the pruner after being enabled for a while, the history is already incomplete, but with the current design we will assume we have complete history -- can we always initialize the pruners, but optionally enable the workers depending on config? This way we always know the min_readable_version. Also avoids a lot of `if Some(pruner) = ..`

Reply:
1. `error_if_version_is_pruned` will take either ledger pruner or state pruner so want to make it generic.
2. Can we do it in a separate PR after this refactoring PR?

Reply:
1. `fn fun(ref: &dyn Tr)` will work `fun(&ty)` if ty implements `Tr`

Reply:
Addressed 1. Will have a follow up PR to do 2.